testcases/crypto/abfunc.c

```diff
@@ -0,0 +1,1079 @@
+                                &handle);
+        if (rc != config[i].exprc) {
+            testcase_fail("Derive Test %lu got unexpected derive result (got: %s; expected %s)",
+                          i, p11_get_ckr(rc), p11_get_ckr(config[i].exprc));
+            continue;
+        }
+        if (rc == CKR_OK) {
+            rc = funcs->C_GetAttributeValue(session, handle, &checkattr, 1);
+            if (rc == CKR_TEMPLATE_INCOMPLETE) {
+                // Treat this as okay, but non-AB
+                rc = CKR_OK;
+                checkval = FALSE;
+            }
+            if (rc != CKR_OK) {
+                testcase_fail("Derive Test %lu failed to retrieve AB attribute (rc=%s)",
+                              i, p11_get_ckr(rc));
+            } else if (checkval == config[i].expabval) {
+                testcase_pass("Derive Test %lu: cycle completed", i);
+            } else {
+                testcase_fail("Derive Test %lu got wrong AB value (expected %s)",
+                              i, config[i].expabval ? "TRUE" : "FALSE");
+            }
+        } else {
+            testcase_pass("Derive Test %lu failed as expected", i);
+        }
+        funcs->C_DestroyObject(session, handle);
+    }
+}
+
+void testdriver()
+{
+    CK_BYTE user_pin[PKCS11_MAX_PIN_LEN];
+    CK_ULONG user_pin_len;
+    CK_SESSION_HANDLE session;
+    CK_FLAGS flags;
+    CK_RV rc = CKR_OK;
+
+    testsuite_begin("Attribute Bound Keys Tests");
+    testcase_rw_session();
+    testcase_user_login();
+
+    do_TestInvalidKeys(session);
+
+    do_CreateKeyObjects(session);
+
+    rc = do_SetupKeys(session);
+    if (rc != CKR_OK) {
+        testcase_error("Bail out since keys were not properly created!");
+        goto testcase_cleanup;
+    }
+    do_TestKeyWrappingUnwrapping(session);
+    do_TestDerive(session);
+    do_TestAttributeChange(session);
+ testcase_cleanup:
+    teardown(session);
+    testcase_user_logout();
+    testcase_close_session();
+}
+
+int main(int argc, char **argv)
+{
+    int rc;
+    CK_C_INITIALIZE_ARGS cinit_args;
+    CK_RV rv;
+
+    rc = do_ParseArgs(argc, argv);
+    if (rc != 1) {
+        return rc;
+    }
+
+    printf("Using slot #%lu...\n\n", SLOT_ID);
+    printf("With option: no_stop: %d\n", no_stop);
+
+    rc = do_GetFunctionList();
+    if (!rc) {
+        PRINT_ERR("ERROR do_GetFunctionList() Failed, rx = 0x%0x\n", rc);
+        return rc;
+    }
+
+    memset(&cinit_args, 0x0, sizeof(cinit_args));
+    cinit_args.flags = CKF_OS_LOCKING_OK;
+
+    funcs->C_Initialize(&cinit_args);
+    {
+        CK_SESSION_HANDLE hsess = 0;
+        rc = funcs->C_GetFunctionStatus(hsess);
+        if (rc != CKR_FUNCTION_NOT_PARALLEL) {
+            return rc;
+        }
+
+        rc = funcs->C_CancelFunction(hsess);
+        if (rc != CKR_FUNCTION_NOT_PARALLEL) {
+            return rc;
+        }
+    }
+
+    if (is_ep11_token(SLOT_ID)) {
+        testcase_setup(0);
+        rv = do_CheckMechanismInfo();
```

Comment:
So if CKM_IBM_ATTRIBUTEBOUND_WRAP is not supported, then you continue anyway? Shouldn't you skip the call to testdriver then? 
You could also omit the check for is_ep11_token then, since other tokens simply wont support CKM_IBM_ATTRIBUTEBOUND_WRAP and thus skip the tests.

Reply:
Well, skipping the call to testdriver was certainly my intention...  Fixed.